v0/destinations/mailchimp/transform.js

```diff
@@ -1,107 +1,153 @@
+    if (email) {
+      allMergedFields = mergeAdditionalTraitsFields(traits, mergedFieldPayload);
+    }
+  }
+
+  primaryPayload.merge_fields = allMergedFields;
+  if (isDefinedAndNotNull(updateSubscription) && emailExists) {
+    Object.keys(updateSubscription).forEach(field => {
+      if (field === "subscriptionStatus") {
+        primaryPayload.status = updateSubscription[field];
+      } else {
+        primaryPayload[field] = updateSubscription[field];
       }
     });
-  } catch (error) {
+  } else if (!emailExists) {
+    const isDoubleOptin = await checkIfDoubleOptIn(
+      apiKey,
+      datacenterId,
+      audienceId
+    );
+    primaryPayload.status = isDoubleOptin
+      ? SUBSCRIPTION_STATUS.pending
+      : SUBSCRIPTION_STATUS.subscribed;
+  }
+  if (
+    primaryPayload.status &&
+    !VALID_STATUSES.includes(primaryPayload.status)
+  ) {
     throw new CustomError(
-      "User does not have access to the requested operation",
-      error.status || 400
+      "The status must be one of [subscribed, unsubscribed, cleaned, pending, transactional]",
+      400
     );
   }
-  return !!response.data.double_optin;
-}
-// New User - make a post request to create the user with the userObj and api key
-function getSubscribeUserUrl(mailChimpConfig) {
-  return `${getMailChimpEndpoint(mailChimpConfig)}/members`;
-}
-
-// Existing user - make a put request to create the user with the userObj and api key
-function getUpdateUserTraitsUrl(mailChimpConfig, email) {
-  const hash = md5(email);
-  return `${getMailChimpEndpoint(mailChimpConfig)}/members/${hash}`;
-}
-
-async function responseBuilderSimple(payload, message, mailChimpConfig) {
-  const email = getFieldValueFromMessage(message, "email");
-  const emailExists = await checkIfMailExists(mailChimpConfig, email);
 
-  const response = defaultRequestConfig();
+  return removeUndefinedAndNullValues(primaryPayload);
+};
+const formFinalEndPoint = (datacenterId, audienceId, emailExists, email) => {
+  let endpoint;
   if (emailExists) {
-    response.endpoint = getUpdateUserTraitsUrl(mailChimpConfig, email);
-    response.method = defaultPutRequestConfig.requestMethod;
+    const hash = md5(email);
+    endpoint = `${getMailChimpEndpoint(
+      datacenterId,
+      audienceId
+    )}/members/${hash}`;
   } else {
-    response.endpoint = getSubscribeUserUrl(mailChimpConfig);
-    response.method = defaultPostRequestConfig.requestMethod;
+    endpoint = `${getMailChimpEndpoint(datacenterId, audienceId)}/members`;
   }
-  response.body.JSON = payload;
-  const basicAuth = Buffer.from(`apiKey:${mailChimpConfig.apiKey}`).toString(
-    "base64"
+  return endpoint;
+};
+const responseBuilderSimple = async (
+  finalPayload,
+  message,
+  messageConfig,
+  audienceId,
+  emailExists
+) => {
+  const { datacenterId, apiKey } = messageConfig;
+  const response = defaultRequestConfig();
+  const email = getFieldValueFromMessage(message, "email");
+  response.endpoint = formFinalEndPoint(
```

Comment:
Lets use separate functions for case when email exists and it doesn't ref: https://softwareengineering.stackexchange.com/questions/147977/is-it-wrong-to-use-a-boolean-parameter-to-determine-behavior

Reply:
done